Port AMSI scanning for assembly loads

elinor-fung opened the pull request:
Pass all assemblies loaded from byte arrays to AmsiScanBuffer on Windows.
Port from .NET Framework 4.8

Fixes #21370

jeffschwMSFT commented on src/vm/peimagelayout.cpp:
```diff
@@ -3,12 +3,16 @@
 #include "peimagelayout.inl"
 
+#if !defined(PLATFORM_UNIX) && !defined(CROSSGEN_COMPILE)
+#include "amsi.h"
```
Style thing (and I am not sure of the answer) do we usual do excludes for predefines?  Or add includes.  In this case it would be Win32 and Win64 only.

elinor-fung replied:
@jkotas would you be the one with the answer? Is there a standard style we prefer here?

jkotas replied:
We do not have 100% uniform style for this. I think the most common style for this is `#if !defined(FEATURE_PAL)`, but `#if !defined(PLATFORM_UNIX)` is used as well.

I think it would be more readable as `#if defined(PLATFORM_WINDOWS) ...`, but we do not have `PLATFORM_WINDOWS` defined for Windows build. Feel free to add to https://github.com/dotnet/coreclr/blob/master/configurecompiler.cmake if you would like. It is confusing that we do not have `PLATFORM_WINDOWS` defined for Windows.

When everything else is equal, the code with no or fewer negated conditions is better for readability.

elinor-fung replied:
Updating to define/use `PLATFORM_WINDOWS`

elinor-fung commented:
@dotnet-bot help

AaronRobinsonMSFT commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test please

elinor-fung commented:
@dotnet-bot  test Ubuntu x64 Checked CoreFX Tests

GrabYourPitchforks commented on src/vm/amsi.cpp:
```diff
@@ -0,0 +1,147 @@
+        return true;
+    }
+
+    bool HasLoadLibrarySearchSystem32()
```
There's no need to test for OS support for KB2533623 or LOAD_LIBRARY_SEARCH_SYSTEM32. The patch has been out for 8½ years. There's no in-market operating system under active support that doesn't already include the patch as part of its base image.

elinor-fung replied:
Ah, thanks. Removed.

GrabYourPitchforks commented on src/vm/amsi.cpp:
```diff
@@ -0,0 +1,147 @@
+        return false;
+
+    DWORD result;
+    HRESULT hr = AmsiScanBuffer(s_amsiContext, flatImageBytes, size, nullptr, nullptr, &result);
```
What other HRESULTs can this return? Should we assume the payload is ok if `AmsiScanBuffer` returns `E_SOMETHING`?

elinor-fung replied:
It only returns S_OK if it successfully scanned and anything else indicates it did not successfully scan (none of the AMSI providers succeeded). In that case, we treat it the same as if AMSI isn't available and do not block loading the bytes payload (matching full Framework behaviour).

GrabYourPitchforks replied:
Sounds good - thanks for the clarification! :)

adityamandaleeka commented:
@GrabYourPitchforks I'll leave the decision on whether to keep/remove the whitespace changes to @elinor-fung, but here's a GitHub tip for future reference: you can add `?w=1` to the end of a diff URL to ignore whitespace changes.

elinor-fung commented:
Updated for nicer predefine (added `PLATFORM_WINDOWS`) and removed the `LOAD_LIBRARY_SEARCH_SYSTEM32` support check.

I haven't reverted the whitespace changes (but I did set my editor to not trim trailing whitespace anymore). If it is really a big problem, I can go through and revert them tomorrow.

jkotas commented:
> I haven't reverted the whitespace changes

I think that's fine.

GrabYourPitchforks commented:
> If it is really a big problem, I can go through and revert them tomorrow.

It's not a huge deal to be honest. It just makes reviewing / isolating the change a bit difficult because of additional noise. But it's really just a minor complaint in the grand scheme of things.

elinor-fung commented:
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Release crossgen_comparison Build and Test

GrabYourPitchforks commented:
Thanks so much @elinor-fung for getting to this! Much appreciated. :)
